plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,833 @@
+    output_directory : `~pathlib.Path`
+        Output for objects that are saved to disk
+
+
+    Notes
+    -----
+    After every push, the `post_push_hook` method is called with the
+    respective `~plasmapy.simulation.particle_tracker.ParticleTracker` object passed as a parameter.
+    Then, the hook calls `save_now` to determine whether or not the simulation state should be saved.
+    If it is determined that the simulation state should be saved, the save routine may call either
+    `save_to_disk` or `save_to_memory` depending on the routine implemented.
+    """
+
+    @property
+    def particle_tracker(self):
+        """Return the `ParticleTracker` object for this stop condition."""
+        return self._particle_tracker
+
+    @particle_tracker.setter
+    def particle_tracker(self, particle_tracker):
+        self._particle_tracker = particle_tracker
+
+    @property
+    @abstractmethod
+    def require_uniform_dt(self):
+        """Return whether or not this save routine requires a uniform dt to be specified."""
+        ...
+
+    @property
+    @abstractmethod
+    def save_now(self):
+        """Determine whether or not to save on the current push step."""
+        ...
+
+    @abstractmethod
+    def save(self):
+        """The abstract method for saving the current state of the |ParticleTracker|."""
+        ...
+
+    def post_push_hook(self, force_save=False):
+        """Function called after a push step.
+
+        This function is responsible for handling two steps of save routine, namely:
+            - Deciding whether or not to save on the current time step
+            - How the simulation data is saved (i.e. to disk or memory)
+
+        """
+        if self.save_now or force_save:
+            self.save()
+
+
+class AbstractIntervalSaveRoutine(AbstractSaveRoutine, ABC):
+    """Abstract class describing a save routine that saves every given interval."""
+
+    def __init__(self, interval: u.Quantity):
+        self.save_interval = interval.to(u.s).value
+        self.time_of_last_save = 0
+
+    @property
+    def require_uniform_dt(self):
+        """Save output only makes sense for uniform time steps,
+        therefore this routine requires a uniform time step.
+        """
+        return True
+
+    @property
+    def save_now(self):
+        """Save at every interval given in instantiation."""
+        saves_per_step = np.floor(self.save_interval / self._particle_tracker.dt)
+
+        # If the user is requesting multiple saves per step then raise a ValueError
+        if saves_per_step == 0:
+            raise ValueError(
+                "You must specify a time step smaller than the save interval!"
+            )
+
+        if self._particle_tracker.time - self.time_of_last_save >= self.save_interval:
+            self.time_of_last_save = self._particle_tracker.time
+
+            return True
+        else:
+            return False
+
+
+class AbstractDiskSaveRoutine(AbstractSaveRoutine, ABC):
```

Comment:
Is there a reason to not just have  `save_to_memory` and `save_to_disk` methods on the abstract save class, then chose which one to call based on a switch variable? I'm thinking that the way you have it now, you'll have to implement two classes for each type of save condtion (e.g. a save to memory class for interval saving, then another one for saving to disk with the same `save_now` function?

Reply:
I think the issue is then how to deal with the output path. Before, the abstract stopping condition class took an argument for the output path, which doesn't apply when saving to memory. I transferred the output directory component of the code to its own class for the sake of being explicit that the output path parameter is only relevant to disk saving routines.

It would be a good idea to implement the `save_to_memory` and `save_to_disk` methods in the abstract class. My only concern would be that the `save_to_disk` method would have to make reference to some sort of `output_path` parameter, which I hesitate to include in the abstract constructor. Maybe this isn't a problem as long as the documentation is clear?

Reply:
Ok, yes I see. In this case I think it's better to have 'output_path' be an optional keyword to the abstract class, then just note in the documentation that it only needs to be specified if saving to disk. `output_dir` might be better, since you should probably save each timestep to a different h5 file. You can add a quick check that `self.output_diri is not None` to the save to disk method to verify it was set.